File: python/examples/example_using_P0Y0_compair.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import pyrenn as prn

###
#Read Example Data
df = pd.ExcelFile('example_data.xlsx').parse('compressed_air')
P = np.array([df['P1'].values,df['P2'].values,df['P3'].values])
Y = np.array([df['Y1'].values,df['Y2']])
Ptest_ = np.array([df['P1test'].values,df['P2test'].values,df['P3test'].values])
Ytest_ = np.array([df['Y1test'].values,df['Y2test']])

#define the first timestep t=0 of Test Data as previous (known) data P0test and Y0test
P0test = Ptest_[:,0:1]
Y0test = Ytest_[:,0:1]
#Use the timesteps t = [1..99] as Test Data
Ptest = Ptest_[:,1:100]
Ytest = Ytest_[:,1:100]

###
#Create and train NN

#create feed forward neural network with 1 input, 2 hidden layers with 
#4 neurons each and 1 output
#the NN has a recurrent connection with delay of 1 timesteps from the output
# to the first layer
net = prn.CreateNN([3,5,5,2],dIn=[0],dIntern=[],dOut=[1])


#Train NN with training data P=input and Y=target
#Set maximum number of iterations k_max to 500
#Set termination condition for Error E_stop to 1e-5
#The Training will stop after 500 iterations or when the Error <=E_stop
prn.train_LM(P,Y,net,verbose=True,k_max=500,E_stop=1e-5) 


###
#Calculate outputs of the trained NN for test data with and without previous input P0 and output Y0
ytest = prn.NNOut(Ptest,net)
y0test = prn.NNOut(Ptest,net,P0=P0test,Y0=Y0test)


###
#Plot results
fig = plt.figure(figsize=(15,10))
ax0 = fig.add_subplot(211)
ax1 = fig.add_subplot(212,sharey=ax0)
fs=18

t = np.arange(0,np.shape(Ptest)[1])/4.0 #timesteps in 15 Minute resolution

#Test Data
ax0.set_title('Test Data',fontsize=fs)
ax0.plot(t,ytest[0],color='b',lw=2,label='NN Output without P0,Y0')
ax0.plot(t,y0test[0],color='g',lw=2,label='NN Output with P0,Y0')
ax0.plot(t,Ytest[0],color='r',marker='None',linestyle=':',lw=3,markersize=8,label='Test Data')
ax0.tick_params(labelsize=fs-2)
ax0.legend(fontsize=fs-2,loc='upper right')
ax0.grid()
plt.setp(ax0.get_xticklabels(), visible=False)
ax0.set_ylabel('Storage Pressure [bar]',fontsize=fs)

ax1.plot(t,ytest[1],color='b',lw=2,label='NN Output without P0,Y0')
ax1.plot(t,y0test[1],color='g',lw=2,label='NN Output with P0,Y0')
ax1.plot(t,Ytest[1],color='r',marker='None',linestyle=':',lw=3,markersize=8,label='Test Data')
ax1.tick_params(labelsize=fs-2)
ax1.grid()
ax1.set_xlabel('Time [h]',fontsize=fs)
ax1.set_ylabel('el. Power [kW]',fontsize=fs)

fig.tight_layout()
plt.show()

File: matlab/examples/example_using_P0Y0_compair.m
%%
%Read Example Data
file = 'example_data.xlsx';
num = xlsread(file,'compressed_air');
%Inputs and outputs have to be matrices where columns=datapoints
%and rows=inputs
P = num(:,2:4).';
Y = num(:,5:6).';
Ptest_ = num(:,7:9).';
Ytest_ = num(:,10:11).';

%define the first timestep t=1 of Test Data as previous (known)
%data P0test and Y0test
P0test = Ptest_(:,1);
Y0test = Ytest_(:,1);
%Use the timesteps t = [2..100] as Test Data
Ptest = Ptest_(:,2:100);
Ytest = Ytest_(:,2:100);

%%
%Create NN

%create recurrent neural network with 3 inputs, 2 hidden layers with 
%5 neurons each and 2 outputs
%the NN uses the input data at timestep t-1 and t-2
%The NN has a recurrent connection with delay of 1,2 and 3 timesteps from the output
% to the first layer (and no recurrent connection of the hidden layers)
nn = [3 5 5 2];
dIn = [0];
dIntern=[];
dOut=[1];
net = CreateNN(nn,dIn,dIntern,dOut); %alternative: net = CreateNN([3,5,5,2],[0],[],[1]);

%%
%Train with LM-Algorithm
% Train NN with training data P=input and Y=target
% Set maximum number of iterations k_max to 500
% Set termination condition for Error E_stop to 1e-5
% The Training will stop after 500 iterations or when the Error <=E_stop
net = train_LM(P,Y,net,500,1e-5);


%Calculate Output of trained net (LM) for and Test with and without P0 and Y0
ytest = NNOut(Ptest,net); 
y0test = NNOut(Ptest,net,P0test,Y0test); 

%%
%Plot Results
t = (1:size(Ptest,2))./4; %timesteps in 15 Minute resolution

fig = figure();
set(fig, 'Units', 'normalized', 'Position', [0.2, 0.1, 0.6, 0.6])

subplot(211)
title('Train Data')
set(gca,'FontSize',16)
plot(t,Ytest(1,:),'r:','LineWidth',2)
hold on
grid on
plot(t,ytest(1,:),'b','LineWidth',2)
plot(t,y0test(1,:),'g','LineWidth',2)
l2 = legend('Test Data','NN output without P0,Y0','NN Output with P0,Y0','Location','northoutside','Orientation','horizontal');
set(l2,'FontSize',14)
set(gca, 'LooseInset', get(gca,'TightInset'))
axis tight
ylabel('Storage Pressure [bar]')


subplot(212)
set(gca,'FontSize',16)
plot(t,Ytest(2,:),'r:','LineWidth',2)
hold on
grid on
plot(t,ytest(2,:),'b','LineWidth',2)
plot(t,y0test(2,:),'g','LineWidth',2)
xlabel('time [h]')
set(gca, 'LooseInset', get(gca,'TightInset'))
axis tight
ylabel('el. Power [kW]')
